### PR TITLE
add xmlparser project

### PR DESCRIPTION
**Pull request by OksanaBisyk-TomTom**
dev
## JIRA

* [Main JIRA ticket](https://jira.softserve.academy/secure/RapidBoard.jspa?rapidView=id)

## Summary of change

ToDo needs to add logger instead of cout

### CODE REVIEW

File: XMLParser/XMLParser.h [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+	virtual std::string GetServerName() = 0;
+	virtual std::string GetServerDisplayName() = 0;
+	virtual unsigned int GetListenerPort() = 0;
+	virtual std::string GetIpAddress() = 0;
+	virtual bool UseBlockingSockets() = 0;
+	virtual unsigned int GetSocketTimeOut() = 0;
+	virtual std::string GetLogFilename() = 0;
+	virtual unsigned int GetLogLevel() = 0;
+	virtual bool UseLogFlush() = 0;
+	virtual unsigned int GetPeriodTime() = 0;
```
Comment (CharlesLyleLaRue) [inline XMLParser/XMLParser.h]: Maybe change method name to "GetThreadInterval()".

File: XMLParser/XMLParser.h [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	struct XMLNode
+	{
+		std::string tag;
+		std::string value;
+		std::vector<std::shared_ptr<XMLNode>> children;
+		std::shared_ptr<XMLNode> parent;
+	};
+
+	XMLParser() : m_root{ nullptr }
+	{
```
Comment (CharlesLyleLaRue) [inline XMLParser/XMLParser.h:66]: ?

File: XMLParser/XMLParser.h [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+	}
+
+	~XMLParser()
+	{
+		m_root = nullptr;
```
Comment (CharlesLyleLaRue) [inline XMLParser/XMLParser.h:71]: We don't need to nullify smart pointer.

File: XMLParser/XMLParser.h [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+	}
+
+private:
+	bool ParseFile(const std::string& filename);
+	bool CreateElement(std::shared_ptr<XMLNode>& elem_pointer, const std::string& elem_tag, const std::string& elem_data);
+	bool ChildAppend(std::shared_ptr<XMLNode>& parent, std::shared_ptr<XMLNode>& child);
+	bool FileCheck(const std::filesystem::path& filename) const;
```
Comment (CharlesLyleLaRue) [inline XMLParser/XMLParser.h]: const arg and const method?
Check if that's okay =)

File: XMLParser/XMLParser.h [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+	void FindByTag(const std::shared_ptr<XMLNode>& current_node, const std::string& tag_needed, std::string& data);
```
Comment (CharlesLyleLaRue) [inline XMLParser/XMLParser.h]: Why is that method is not const if it doesn't change member of the class?

File: XMLParser/XMLParser.cpp [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+
+#include "pch.h"
+#include "framework.h"
+#include "XMLParser.h"
+
+bool XMLParser::ParseFile(const std::string& filename)
+{
+	struct Characters
+	{
+		const char tag1 = '<';
```
Comment (CharlesLyleLaRue) [inline XMLParser/XMLParser.cpp]: Const variables should be capitalized.

File: XMLParser/XMLParser.cpp [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+	std::filesystem::path file_path = filename;
+	if (!FileCheck(file_path))
+	{
+		return false;
+	}
+
+	std::ifstream file{ filename };
+	if (!file.is_open())
+	{
+		//std::cout << "Can't open file.\n";
```
Comment (CharlesLyleLaRue) [inline XMLParser/XMLParser.cpp]: Don't forget to change it for logger functions.

File: XMLParser/XMLParser.cpp [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+		return false;
+	}
+
+	std::string line;
+	std::unique_ptr<Characters> symbol = std::make_unique<Characters>();
+	bool tag_flag = false;
+	bool first_flag = true;
+	std::shared_ptr<XMLNode> current_node = nullptr;
```
Comment (CharlesLyleLaRue) [inline XMLParser/XMLParser.cpp]: We don't need to nullify smart pointers.

File: XMLParser/XMLParser.cpp [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+		std::size_t tag_close;
+		if (tag_begin == std::string::npos || tag_end == std::string::npos)
+		{
+			//std::cout << "XML file is not valid.\n";
+			return false;
+		}
+		std::string element_tag;
+		std::string element_data;
+		std::string close_tag;
+		int one = 1;
```
Comment (CharlesLyleLaRue) [inline XMLParser/XMLParser.cpp]: It has to be better names for this variables (one, two).

File: XMLParser/XMLParser.cpp [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+			parent_node = current_node;
+			first_flag = false;
+		}
+		else
+		{
+			ChildAppend(parent_node, current_node);
+			if (tag_close == std::string::npos)
+			{
+				parent_node = current_node;
+				current_node = nullptr;
```
Comment (CharlesLyleLaRue) [inline XMLParser/XMLParser.cpp:98]: It's smart pointer, we should use ptr.reset().

File: XMLParser/XMLParser.cpp [hunk 11]
```diff
@@ -0,0 +1,361 @@
...
+		return false;
+	}
+	parent->children.push_back(child);
+	child->parent = parent;
+	return true;
+}
+
+bool XMLParser::FileCheck(const std::filesystem::path& filename) const
+{
+	if (!std::filesystem::exists(filename))
```
Comment (CharlesLyleLaRue) [inline XMLParser/XMLParser.cpp:131]: Replace cout for logger functions, don't forget.

File: XMLParser/XMLParser.cpp [hunk 12]
```diff
@@ -0,0 +1,361 @@
...
+		else
+		{
+			FindByTag(current_node->children[i], tag_needed, data);
+		}
+	}
+}
+
+bool XMLParser::ValueCheck(const std::string& line, unsigned int& number)
+{
+	std::string data = "";
```
Comment (CharlesLyleLaRue) [inline XMLParser/XMLParser.cpp]: We don't need to initialize string value.

File: XMLParser/XMLParser.cpp [hunk 13]
```diff
@@ -0,0 +1,361 @@
...
+}
+
+std::string XMLParser::GetServerName()
+{
+	if (m_root == nullptr)
+	{
+		ParseFile(XML_FILE_PATH);
+	}
+	std::string line = "servername";
+	std::string data = "";
```
Comment (CharlesLyleLaRue) [inline XMLParser/XMLParser.cpp]: We don't need to initialize string value.

File: XMLParser/XMLParser.cpp [hunk 14]
```diff
@@ -0,0 +1,361 @@
...
+}
+
+std::string XMLParser::GetServerDisplayName()
+{
+	if (m_root == nullptr)
+	{
+		ParseFile(XML_FILE_PATH);
+	}
+	std::string line = "serverdisplayname";
+	std::string data = "";
```
Comment (CharlesLyleLaRue) [inline XMLParser/XMLParser.cpp]: We don't need to initialize string value.

File: XMLParser/XMLParser.cpp [hunk 15]
```diff
@@ -0,0 +1,361 @@
...
+}
+
+unsigned int XMLParser::GetListenerPort()
+{
+	if (m_root == nullptr)
+	{
+		ParseFile(XML_FILE_PATH);
+	}
+	std::string line = "listenerport";
+	unsigned int port = 0;
```
Comment (CharlesLyleLaRue) [inline XMLParser/XMLParser.cpp:231]: Do we really need to initialize this value?

File: XMLParser/XMLParser.cpp [hunk 16]
```diff
@@ -0,0 +1,361 @@
...
+}
+
+std::string XMLParser::GetIpAddress()
+{
+	if (m_root == nullptr)
+	{
+		ParseFile(XML_FILE_PATH);
+	}
+	std::string line = "ipaddress";
+	std::string data = "";
```
Comment (CharlesLyleLaRue) [inline XMLParser/XMLParser.cpp]: We don't need to initialize string value.

File: XMLParser/XMLParser.cpp [hunk 17]
```diff
@@ -0,0 +1,361 @@
...
+}
+
+std::string XMLParser::GetLogFilename()
+{
+	if (m_root == nullptr)
+	{
+		ParseFile(XML_FILE_PATH);
+	}
+	std::string line = "filename";
+	std::string data = "";
```
Comment (CharlesLyleLaRue) [inline XMLParser/XMLParser.cpp]: We don't need to initialize string value.

File: XMLParser/XMLParser.cpp [hunk 18]
```diff
@@ -0,0 +1,361 @@
...
+}
+
+bool XMLParser::UseBlockingSockets()
+{
+	if (m_root == nullptr)
+	{
+		ParseFile(XML_FILE_PATH);
+	}
+	std::string line = "blocking";
+	unsigned int block = 0;
```
Comment (CharlesLyleLaRue) [inline XMLParser/XMLParser.cpp:261]: We don't need to initialize this value.

File: XMLParser/XMLParser.cpp [hunk 19]
```diff
@@ -0,0 +1,361 @@
...
+}
+
+unsigned int XMLParser::GetSocketTimeOut()
+{
+	if (m_root == nullptr)
+	{
+		ParseFile(XML_FILE_PATH);
+	}
+	std::string line = "socket_timeout";
+	unsigned int time_out = 0;
```
Comment (CharlesLyleLaRue) [inline XMLParser/XMLParser.cpp:272]: We don't need to initialize this value.

File: XMLParser/XMLParser.cpp [hunk 20]
```diff
@@ -0,0 +1,361 @@
...
+}
+
+unsigned int XMLParser::GetLogLevel()
+{
+	if (m_root == nullptr)
+	{
+		ParseFile(XML_FILE_PATH);
+	}
+	std::string line = "LogLevel";
+	unsigned int log_level = 0;
```
Comment (CharlesLyleLaRue) [inline XMLParser/XMLParser.cpp:302]: We don't need to initialize this value.

File: XMLParser/XMLParser.cpp [hunk 21]
```diff
@@ -0,0 +1,361 @@
...
+}
+
+bool XMLParser::UseLogFlush()
+{
+	if (m_root == nullptr)
+	{
+		ParseFile(XML_FILE_PATH);
+	}
+	std::string line = "flush";
+	unsigned int flush = 0;
```
Comment (CharlesLyleLaRue) [inline XMLParser/XMLParser.cpp:317]: We don't need to initialize this value.

File: XMLParser/XMLParser.cpp [hunk 22]
```diff
@@ -0,0 +1,361 @@
...
+}
+
+unsigned int XMLParser::GetPeriodTime()
+{
+	if (m_root == nullptr)
+	{
+		ParseFile(XML_FILE_PATH);
+	}
+	std::string line = "Period_time";
+	unsigned int time = 0;
```
Comment (CharlesLyleLaRue) [inline XMLParser/XMLParser.cpp:328]: We don't need to initialize this value.

File: XMLParser/XMLParser.cpp [hunk 23]
```diff
@@ -0,0 +1,361 @@
...
+}
+
+unsigned int XMLParser::GetMaxWorkingThreads()
+{
+	if (m_root == nullptr)
+	{
+		ParseFile(XML_FILE_PATH);
+	}
+	std::string line = "maxworkingthreads";
+	unsigned int number = 0;
```
Comment (CharlesLyleLaRue) [inline XMLParser/XMLParser.cpp:343]: We don't need to initialize this value.

**Review by CharlesLyleLaRue — CHANGES_REQUESTED** [review]
Please, check my comments.

File: XMLParser/XMLParser.cpp [hunk 24]
```diff
@@ -0,0 +1,361 @@
...
+		//std::cout << "Can't open file.\n";
+		return false;
+	}
+
+	std::string line;
+	std::unique_ptr<Characters> symbol = std::make_unique<Characters>();
+	bool tag_flag = false;
+	bool first_flag = true;
+	std::shared_ptr<XMLNode> current_node = nullptr;
+	std::shared_ptr<XMLNode> parent_node = nullptr;
```
Comment (KostiantynYehorov) [inline XMLParser/XMLParser.cpp]: We don't need to set out "parent-node" to "nullptr"

**Review by KostiantynYehorov — DISMISSED** [review]
Good, clean code)

File: XMLParser/XMLParser.h (hunk 5 above)
Comment (OksanaBisyk-TomTom) [reply XMLParser/XMLParser.h]: Done.

File: XMLParser/XMLParser.cpp (hunk 7 above)
Comment (OksanaBisyk-TomTom) [reply XMLParser/XMLParser.cpp]: Yes, sir.

File: XMLParser/XMLParser.cpp (hunk 8 above)
Comment (OksanaBisyk-TomTom) [reply XMLParser/XMLParser.cpp]: We don't need to, but it's better for code readability.

File: XMLParser/XMLParser.cpp (hunk 9 above)
Comment (OksanaBisyk-TomTom) [reply XMLParser/XMLParser.cpp]: Renamed.

File: XMLParser/XMLParser.cpp (hunk 10 above)
Comment (OksanaBisyk-TomTom) [reply XMLParser/XMLParser.cpp:98]: It's a matter of code readability.

File: XMLParser/XMLParser.cpp (hunk 13 above)
Comment (OksanaBisyk-TomTom) [reply XMLParser/XMLParser.cpp]: Done.

File: XMLParser/XMLParser.cpp (hunk 24 above)
Comment (OksanaBisyk-TomTom) [reply XMLParser/XMLParser.cpp]: We don't need to, but it's better for code readability.

**Review by CharlesLyleLaRue — DISMISSED** [review]
Looks good.

**Review by KostiantynYehorov — DISMISSED** [review]

**Pull request closed without merging**